File: src/main/java/io/github/shaksternano/mediamanipulator/util/GithubUtil.java
package io.github.shaksternano.mediamanipulator.util;

import com.google.common.collect.ImmutableList;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;

import java.util.*;

public class GithubUtil {

    /**
     * Gets the list of file names in the given directory.
     *
     * @param user       The name of the user that owns the repository.
     * @param repository The name of the repository.
     * @param commitSha  The SHA of the commit to get the list of files from.
     * @param path       The path to the directory to get the list of files from.
     * @return The list of file names in the given directory.
     */
    public static List<String> listFiles(String user, String repository, String commitSha, String... path) {
        return listFiles(user, repository, commitSha, new LinkedList<>(Arrays.asList(path)));
    }

    private static List<String> listFiles(String user, String repository, String commitSha, List<String> path) {
        String url = getRepositoryApiUrl(user, repository) + "/git/trees/" + commitSha;
        JsonElement content = NetworkUtil.httpGet(url);
        Optional<JsonElement> treeElementOptional = JsonUtil.getNestedElement(content, "tree");
        Optional<JsonArray> treeArrayOptional = treeElementOptional.map(treeElement -> {
            if (treeElement.isJsonArray()) {
                return treeElement.getAsJsonArray();
            } else {
                return null;
            }
        });

        if (treeArrayOptional.isPresent()) {
            JsonArray treeArray = treeArrayOptional.orElseThrow();
            if (path.isEmpty()) {
                List<String> files = new ArrayList<>();
                for (JsonElement fileElement : treeArray) {
                    JsonUtil.getNestedElement(fileElement, "path").flatMap(JsonUtil::getString).ifPresent(files::add);
                }
                return files;
            } else {
                String firstPath = path.remove(0);
                for (JsonElement treeElement : treeArray) {
                    Optional<String> pathOptional = JsonUtil.getNestedElement(treeElement, "path").flatMap(JsonUtil::getString);
                    if (Optional.of(firstPath).equals(pathOptional)) {
                        Optional<String> shaOptional = JsonUtil.getNestedElement(treeElement, "sha").flatMap(JsonUtil::getString);
                        if (shaOptional.isPresent()) {
                            return listFiles(user, repository, shaOptional.orElseThrow(), path);
                        }
                    }
                }
            }
        }

        return ImmutableList.of();
    }

    public static Optional<String> getLatestReleaseTagCommitSha(String user, String repository) {
        Optional<String> latestReleaseTagOptional = getLatestReleaseTag(user, repository);
        if (latestReleaseTagOptional.isPresent()) {
            JsonElement tags = NetworkUtil.httpGet("https://api.github.com/repos/twitter/twemoji/tags");
            if (tags.isJsonArray()) {
                JsonArray tagsArray = tags.getAsJsonArray();
                for (JsonElement tagElement : tagsArray) {
                    Optional<String> tag = JsonUtil.getNestedElement(tagElement, "name").flatMap(JsonUtil::getString);
                    if (latestReleaseTagOptional.equals(tag)) {
                        return JsonUtil.getNestedElement(tagElement, "commit", "sha").flatMap(JsonUtil::getString);
                    }
                }
            }
        }

        return Optional.empty();
    }

    private static Optional<String> getLatestReleaseTag(String user, String repository) {
        String url = getRepositoryApiUrl(user, repository) + "/releases/latest";
        JsonElement latestRelease = NetworkUtil.httpGet(url);
        return JsonUtil.getNestedElement(latestRelease, "tag_name").flatMap(JsonUtil::getString);
    }

    private static String getRepositoryApiUrl(String user, String repository) {
        return "https://api.github.com/repos/" + user + "/" + repository;
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/util/GitHubUtil.kt
package io.github.shaksternano.borgar.core.util

import io.github.shaksternano.borgar.core.io.httpGet
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Gets the list of file names in the given directory.
 *
 * @param user       The name of the user that owns the repository.
 * @param repository The name of the repository.
 * @param commitSha  The SHA of the commit to get the list of files from.
 * @param path       The path to the directory to get the list of files from.
 * @return The list of file names in the given directory.
 */
suspend fun listGitHubFiles(user: String, repository: String, commitSha: String, vararg path: String): List<String> =
    listGitHubFiles(user, repository, commitSha, path.asList())

private tailrec suspend fun listGitHubFiles(
    user: String,
    repository: String,
    commitSha: String,
    path: List<String>
): List<String> {
    val url = getRepositoryApiUrl(user, repository) + "/git/trees/" + commitSha
    val tree = httpGet<GitHubTreeResponse>(url).tree
    return if (path.isEmpty()) {
        tree.map {
            it.path
        }
    } else {
        val firstPath = path.first()
        val treeDirectory = tree.first {
            it.path == firstPath
        }
        listGitHubFiles(user, repository, treeDirectory.sha, path.drop(1))
    }
}

suspend fun getGitHubLatestReleaseTagCommitSha(user: String, repository: String): String {
    val latestReleaseTag = getLatestReleaseTag(user, repository)
    val tags = httpGet<List<GitHubTag>>("https://api.github.com/repos/twitter/twemoji/tags")
    return tags.first {
        it.name == latestReleaseTag
    }.commit.sha
}

private suspend fun getLatestReleaseTag(user: String, repository: String): String {
    val url = getRepositoryApiUrl(user, repository) + "/releases/latest"
    val latestRelease = httpGet<GitHubReleaseResponse>(url)
    return latestRelease.tagName
}

private fun getRepositoryApiUrl(user: String, repository: String): String =
    "https://api.github.com/repos/$user/$repository"

@Serializable
private data class GitHubReleaseResponse(
    @SerialName("tag_name")
    val tagName: String,
)

@Serializable
private data class GitHubTag(
    val name: String,
    val commit: GitHubCommit,
)

@Serializable
private data class GitHubCommit(
    val sha: String,
)

@Serializable
private data class GitHubTreeResponse(
    val tree: List<GitHubTreeElement>,
)

@Serializable
private data class GitHubTreeElement(
    val path: String,
    val sha: String,
)
